components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java

```diff
@@ -1544,6 +1544,29 @@ public static String getSPTenantDomainFromClientId(String clientId) {
         }
     }
 
+    /**
+     * This method verifies the service provider tenant domain using the client ID.
+     *
+     * @param clientId Client id of the application.
+     * @return tenantDomain domain of the service provider.
+     */
+    public static String verifyAndRetrieveTenantDomain(String clientId)
+            throws OAuthSystemException {
+
+        try {
+            String extractedTenantDomain = OAuth2Util.getLoginTenant();
+            OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientId(clientId, extractedTenantDomain);
+            String appTenantDomain = OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO);
+            if (StringUtils.equals(extractedTenantDomain, appTenantDomain)) {
+                return appTenantDomain;
+            }
+            throw new OAuthSystemException("Provided tenant domain: " + extractedTenantDomain + " does not " +
```

Comment:
Is this a new type of exception? Are we using this in other places? isn't this a client error.

Reply:
We have used this exception in the Endpoint utils